pageserver/src/page_service.rs

```diff
@@ -880,16 +921,148 @@ impl PageServerHandler {
         }))
     }
 
-    async fn do_handle_get_page_at_lsn_request(
-        &self,
-        timeline: &Timeline,
+    /// For most getpage requests, we will already have a Timeline to serve the request: this function
+    /// looks up such a Timeline synchronously and without touching any global state.
+    fn get_cached_timeline_for_page(
+        &mut self,
+        req: &PagestreamGetPageRequest,
+    ) -> Result<&Arc<Timeline>, Key> {
+        let key = if let Some((first_idx, first_timeline)) = self.shard_timelines.iter().next() {
+            // Fastest path: single sharded case
+            if first_idx.shard_count < ShardCount(2) {
+                return Ok(&first_timeline.timeline);
+            }
```

Comment:
Can this become stale during splits?

Reply:
Interesting to think about.

I think the answer is no: the presence of the parent timeline in shard_timelines implies holding a GateGuard on it.  As long as this parent is running, it is okay to send I/O to it, even if we already created some child shards -- there is a period during split where both parent and children are running.

When we shut down the parent:
- we first replace its slot with an InProgress so that future callers to `get_active_tenant_with_timeout` 
- it will shut down any `PageServerHandler` that was using it, the client will reconnect and the new handler will get a fresh cache.

Added a note in https://github.com/neondatabase/neon/issues/6278 to circle back and check all this is true.